pkg/urbit/jets/e/de_json.c

```diff
@@ -0,0 +1,238 @@
+
+  //
+  // initialization
+  //
+
+  // XX assumes little-endian
+  //
+  if ( c3y == u3a_is_cat(txt) ) {
+    byt_y = (c3_y*)&txt;
+  }
+  else {
+    u3a_atom* vat_u = u3a_to_ptr(txt);
+    byt_y = (c3_y*)vat_u->buf_w;
+  }
+  json_open_buffer(sam_u, byt_y, len_w);
+  json_set_allocator(sam_u, &loc_u);
+  u3a_pile_prep(pil_u, sizeof(u3qedj_coll));
+
+  //
+  // core logic
+  //
+
+  while ( json_peek(sam_u) != JSON_DONE ) {
+    switch ( json_next(sam_u) ) {
+      // unreachable barring programming error
+      default: c3_assert(0);
+
+      case JSON_ARRAY:
+      case JSON_OBJECT: {
+        tak_u = _push_stack(pil_u);
+      } continue;
+
+      case JSON_ARRAY_END: {
+        val = u3nc(c3__a, tak_u->col);
+        tak_u = _pop_stack(pil_u);
+      } break;
+
+      case JSON_OBJECT_END: {
+        val = u3nc(c3__o, tak_u->col);
+        tak_u = _pop_stack(pil_u);
+      } break;
+
+      case JSON_STRING: {
+        size_t      len_i;
+        const c3_c *str_c = json_get_string(sam_u, &len_i);
+        u3_atom     str = (len_i <= 1) ?
+                            u3_nul :
+                            u3i_bytes(len_i - 1, (const c3_y *)str_c);
+
+        if ( (json_get_context(sam_u, &cnt_d) == JSON_OBJECT) && (cnt_d & 1) ) {
+          tak_u->key = str;
+        }
+        else {
+          val = u3nc(c3__s, str);
+        }
+      } break;
+
+      case JSON_NUMBER: {
+        size_t      len_i;
+        const c3_c *str_c = json_get_string(sam_u, &len_i);
+
+        if ( len_i > 1 ) {
+          val = u3nc(c3__n, u3i_bytes(len_i - 1, (const c3_y *)str_c));
+        }
+        else {
+          val = u3nc(c3__n, u3_nul);
+        }
+      } break;
+
+      case JSON_TRUE: {
+        val = u3nc(c3__b, c3y);
+      } break;
+
+      case JSON_FALSE: {
+        val = u3nc(c3__b, c3n);
+      } break;
+        
+      case JSON_NULL: {
+        val = u3_nul;
+      } break;
+
+      case JSON_ERROR: {
+        _close(sam_u, pil_u);
+        return u3_nul;
+      } break;
+    }
+
+    switch ( json_get_context(sam_u, &cnt_d) ) {
+      case JSON_DONE: {
+        res = val;
+      } break;
+
+      case JSON_ARRAY: {
+        u3_noun* nex;
+        u3_noun* hed;
+
+        if ( tak_u->col == u3_nul ) {
+          nex = &(tak_u->col);
+        }
+        else {
+          nex = tak_u->tel;
+        }
+
+        *nex = u3i_defcons(&hed, &(tak_u->tel));
+        *hed = val;
+        *(tak_u->tel) = u3_nul;
+      } break;
+
+      case JSON_OBJECT: {
+        if ( !(cnt_d & 1) ) {
+
+          // cnt_d == 0 weeded out by continue command
+          c3_assert(tak_u->key != u3_none);
+
+          tak_u->col = u3kdb_put(tak_u->col, tak_u->key, val);
+          tak_u->key = u3_none;
+        }
+      } break;
+    }
+  }
+
+  //
+  // clean up
+  //
+
+  c3_assert(c3y == u3a_pile_done(pil_u));
+
+  // skip over whitespce
+  while ( json_isspace(json_source_peek(sam_u)) ) {
+    json_source_get(sam_u);
+  }
+
+  _close(sam_u, pil_u);
```

Comment:
`_close_stack()` is a no-op here, demonstrated by the preceding assertion. It might be clearer to just call `json_close()`. The early return in `JSON_ERROR` is the only place where stack "unwinding" is needed.

Reply:
Good point - done